File: _roadmap.py
#!/usr/bin/env python3
"""
_roadmap.py  -  the course roadmap: which stages exist, in what order, and the
markup that draws them. One list, one renderer, three generators.

WHY IT IS AT THE ROOT. The roadmap belongs to no stage. It was declared inside
wireframes/_generators/resync_sidebar.py, which made concept/ have to reach into
the wireframes tree to draw its own panel, a dependency the folder layout denies.
It sits beside the other maps owned by neither layer: _twins.py (which screen is
which screen's twin), _course_chrome.py (what the panel looks like), and
_resync_roadmap.py, which patches the hand-typed copies this file cannot reach.

WHAT IT REPLACED, AND WHAT THAT COST. The list existed four times: LAYOUT here,
and literal HTML typed into ia_annotations.py, into concept/_directions_sidebar.py
and into two stand specimens. Three of the four called themselves the sidebar.
The defect had already fired and had not gone off yet: ia_annotations.py still
held "UI Kit -> ui-kit/kit.html" and "Tokens + Components" as `planned next`,
two stages behind the pages it writes, because _resync_roadmap.py had corrected
the OUTPUT and only one of the two generators. Whoever ran it next would have
rolled fifteen annotation pages back to a roadmap that ended before the stage
they were reading about, and pointed them at a kit frozen as provenance. This is
the pair _levels.py closed between two other generators, in the same shape: one
computation feeds all of its consumers, or they drift and no one can say which
one is lying.

WHAT IS DELIBERATELY NOT HERE. The two stand specimens
(ui-kit/specimens/course-chrome-roadmap.html, ui-kit/specimens.extra.html) show
what the PANEL LOOKS LIKE. Their rows are `href="#"` and abridged on purpose,
and a specimen that re-rendered itself from the truth would start failing when
the truth changed, which is the opposite of what a specimen is for. Each says so
in a comment on its first row.

HOW A CONSUMER DIFFERS FROM ANOTHER. Only in four things, and each is an
argument, so each is a parameter:
  prefix          how deep the page sits (../ or ../../)
  active          which row is the page you are on
  href_override   a row a page can reach without going through the root
  extra_before / extra_after
                  rows this page has and the course does not: the annotation
                  family list, the Concept exploration pages

No em dash.
"""
import os
import sys

ROOT = os.path.dirname(os.path.abspath(__file__))
if ROOT not in sys.path:
    sys.path.insert(0, ROOT)
from _course_chrome import mark_group  # noqa: E402

# ---------------------------------------------------------------------------
# Collapsible GROUPS - stage clusters with more than one sub-page. When the
# active page is inside a group the group renders EXPANDED (its label divider +
# rows). Otherwise it renders COLLAPSED: one link to the group's first page.
#   rows: ("divider", label) | ("item", key, label, href)
# ---------------------------------------------------------------------------
GROUPS = {
    "user-research": {
        "label": "User Research",
        "first": "user-research/personas.html",
        "keys": {"personas", "jtbd", "cjm-as-is", "cjm-to-be"},
        "rows": [
            ("item", "personas", "Personas", "user-research/personas.html"),
            ("item", "jtbd", "JTBD", "user-research/jtbd.html"),
            ("item", "cjm-as-is", "CJM As-Is", "user-research/cjm-as-is.html"),
            ("item", "cjm-to-be", "CJM To-Be", "user-research/cjm-to-be.html"),
        ],
    },
    "ia": {
        "label": "Information Architecture",
        "first": "ia/flows.html",
        "keys": {"flows", "concept-map", "overview", "sitemap", "seo", "system"},
        "rows": [
            ("divider", "Basic layer"),
            ("item", "flows", "Flows", "ia/flows.html"),
            ("item", "concept-map", "Concept map", "ia/concept-map.html"),
            ("divider", "Detailed layer"),
            ("item", "overview", "Overview", "ia/ia.html"),
            ("item", "sitemap", "Sitemap", "ia/sitemap.html"),
            ("item", "seo", "SEO layer", "ia/seo.html"),
            ("item", "system", "System nodes", "ia/system.html"),
        ],
    },
}

# ---------------------------------------------------------------------------
# Top-level roadmap order. Entries:
#   ("item",    key, label, href_or_None, planned_bool)
#   ("group",   group_key)                 # one of GROUPS
#   ("divider", label)                     # standalone section header
# href is relative to repo root; the caller's prefix is prepended at render time.
#
# Twelve stages, thirteen rows: Wireframe Annotations is an artifact of two
# stages rather than a stage of its own, and it earns a row because it is a
# place you can stand. CJM has no row here at all for the opposite reason: it is
# built inside User Research, so it is one of that group's sub-pages.
# ---------------------------------------------------------------------------
LAYOUT = [
    ("item", "foundation", "Foundation Research", "research/research.html", False),
    ("group", "user-research"),
    ("group", "ia"),
    ("divider", "Plan"),
    ("item", "wireframes", "Wireframes", "wireframes/event-feed.html", False),
    ("item", "annotations", "Wireframe Annotations", "ia/annotations/index.html", False),
    ("item", "voice", "Voice", "voice/voice.html", False),
    ("divider", "Design and Delivery"),
    ("item", "concept", "Concept", "concept/concept.html", False),
    ("item", "ui-visual", "UI + Visual", "ui-visual/event-feed.html", False),
    ("item", "tokens-components", "Tokens + Components", "ui-kit/overview.html", False),
    ("item", "design-system", "Design System", None, True),
    ("item", "responsive", "Responsive", None, True),
    ("item", "animation", "Animation", None, True),
    ("item", "handoff", "Handoff", None, True),
]

# The "Next" badge marks the next unbuilt stage: the first planned top-level item.
# If that page ever lives inside a collapsed group, the badge rides the group's
# collapsed link instead (see _render_group).
NEXT_KEY = next((e[1] for e in LAYOUT if e[0] == "item" and e[4]), None)

# The shell every consumer wraps the rows in. The tree is a NAMED <nav>, per
# CLAUDE.md: a panel with no accessible name is a list of links to a screen
# reader, and there are three lists of links on some of these pages.
SHELL_OPEN = [
    '<aside class="sidebar" id="sidebar">',
    '  <div class="sidebar-brand">',
    '    <div class="sidebar-project-name">Prediction Market</div>',
    '  </div>',
    '  <nav class="sidebar-nav" aria-label="Course roadmap">',
]
SHELL_CLOSE = ['  </nav>', '</aside>']


def _render_item(key, label, href, planned, prefix, active, subs_block,
                 href_override, indent):
    """One top-level or in-group row, and its tag says whether it goes anywhere.

    A PLANNED STAGE IS NOT A LINK, so it is not an <a>. It used to be an anchor
    with no href, which is a link element with nothing to open; the badge said
    Soon and the markup said link. A row that goes nowhere is a <span>, and the
    styling never depended on the tag.

    THE ROW YOU ARE ON IS A LINK LIKE EVERY OTHER ROW, which is what the vitrine
    has always done: an active row with no href is a hole in the tree at exactly
    the place a person is standing.
    """
    is_active = key == active
    cls = "sidebar-page-link"
    if is_active:
        cls += " active"
    if planned:
        cls += " planned"
    if key == NEXT_KEY:
        cls += " next"
    if planned:
        out = ['{i}<span class="{cls}">{label}</span>'.format(i=indent, cls=cls, label=label)]
    else:
        target = href_override.get(key)
        if target is None:
            target = prefix + href
        out = ['{i}<a href="{href}" class="{cls}">{label}</a>'.format(
            i=indent, href=target, cls=cls, label=label)]
    if is_active and subs_block:
        out.append(subs_block)
    return out


def _render_group(gkey, prefix, active, subs_block, href_override, indent):
    """Expanded (label divider + rows) when the active page is inside the group,
    otherwise collapsed to one link to the group's first page."""
    g = GROUPS[gkey]
    if active not in g["keys"]:
        cls = "sidebar-page-link"
        if NEXT_KEY in g["keys"]:            # Next rides the collapsed group
            cls += " next"
        return ['{i}<a href="{pfx}{href}" class="{cls}">{label}</a>'.format(
            i=indent, pfx=prefix, href=g["first"], cls=cls, label=g["label"])]
    lines = ['{i}<div class="sidebar-divider">{l}</div>'.format(i=indent, l=g["label"])]
    for row in g["rows"]:
        if row[0] == "divider":
            lines.append('{i}<div class="sidebar-divider">{l}</div>'.format(i=indent, l=row[1]))
        else:
            _, key, label, href = row
            lines.extend(_render_item(key, label, href, False, prefix, active,
                                      subs_block, href_override, indent))
    return lines


def render_rows(prefix, active=None, subs_block="", href_override=None,
                extra_before=None, extra_after=None, indent="    "):
    """Every row of the roadmap, in order, as a list of lines.

    extra_before / extra_after are keyed by the LAYOUT key (or group key) they
    hang off, and their lines are emitted verbatim, so a consumer can add rows
    the course does not have without the course having to know about them.
    """
    href_override = href_override or {}
    extra_before = extra_before or {}
    extra_after = extra_after or {}
    lines = []
    for entry in LAYOUT:
        if entry[0] == "divider":
            lines.append('{i}<div class="sidebar-divider">{l}</div>'.format(i=indent, l=entry[1]))
            continue
        key = entry[1]
        lines.extend(extra_before.get(key, []))
        if entry[0] == "group":
            lines.extend(_render_group(key, prefix, active, subs_block, href_override, indent))
        else:
            _, _, label, href, planned = entry
            lines.extend(_render_item(key, label, href, planned, prefix, active,
                                      subs_block, href_override, indent))
        lines.extend(extra_after.get(key, []))
    return lines


def render_aside(prefix, active=None, **kw):
    """The whole <aside>: brand, named <nav>, every row, and the group you are in
    marked on its label.

    The mark is applied by _course_chrome.mark_group and by nothing else.
    IMPORTED, NOT COPIED: that file rewrites the sixteen course pages these
    generators do not manage, and for one turn each tool undid the other's mark
    and neither reached a fixed point. Two tools may edit one region; they may
    not each have their own idea of what it should say.
    """
    lines = list(SHELL_OPEN) + render_rows(prefix, active, **kw) + list(SHELL_CLOSE)
    return mark_group("\n".join(lines))


def stages():
    """The stages, in course order, as (number, label). Wireframe Annotations is
    a row and not a stage, so it is not numbered."""
    out, n = [], 0
    for e in LAYOUT:
        if e[0] == "group":
            n += 1
            out.append(("%02d" % n, GROUPS[e[1]]["label"]))
        elif e[0] == "item" and e[1] != "annotations":
            n += 1
            out.append(("%02d" % n, e[2]))
    return out


if __name__ == "__main__":
    for num, label in stages():
        print(num, label)

File: assets/_roadmap.js
/* assets/_roadmap.js - the ONE registry of the course roadmap, and the only copy of it.

   WHAT IT REPLACES. The same outline stood as hand-written markup in 28 documents across five
   folders: `research/`, `user-research/`, `ia/` (including `ia/annotations/`), `voice/` and
   `concept/`. 785 lines of it. Turning one planned row into a link on 2026-08-11 was one edit
   repeated 27 times at two path depths, and the 28th file did not need that edit and needed a
   bigger one: `concept/concept.html` stopped at UI + Visual and had stood four rows behind since
   stage 09 shipped. NOTHING NOTICED, BECAUSE A SIDEBAR THAT ENDS EARLY LOOKS LIKE A SIDEBAR, and
   no page declared how long the list was supposed to be. docs/backlog.md 117.

   IT IS THE KIT'S ANSWER, ONE FOLDER OVER. `ui-kit/_nav.js` has held the stand's route since
   2026-08-07 for the same reason and under the same rules, and three of them are taken verbatim:

   1. THE ACTIVE ROW IS COMPUTED FROM THE PATH, never declared. A page cannot carry a stale copy
      of its own name, because the name IS the page. `ui-kit/_nav.js` records what a declared
      `KIT_ACTIVE` cost in the repository it came from.
   2. A ROW WITH NO PAGE IS VISIBLE, as a `<span class="planned">`. Hiding Animation and Handoff
      would make the roadmap look finished and lie about it.
   3. NO CSS OF ITS OWN. Every class written below is `components/course-chrome.css`, which all 28
      pages already link.

   AND ONE THING THAT IS NOT THE KIT'S. The stand's pages are siblings, so its registry holds bare
   file names. These 28 sit at two depths, so every path here is written from the REPOSITORY ROOT
   and the prefix is computed from this script's own `src` at load. A page therefore declares
   neither its name nor its depth, which are the two things a hand-copied panel gets wrong.

   WHAT A PAGE STILL DECLARES, and why it is not drift: its own section anchors, in an
   `<div class="sidebar-sub" data-roadmap-sub>` inside the empty `<aside>`. Those are headings of
   that document and belong to it. The registry never holds them. `ia/annotations/*` is the one
   exception and it runs the other way: its sub list is fourteen SIBLING pages, so it is route and
   it is here.

   THE OUTLINE COLLAPSES, which is why this is a tree and not a list. A stage you are not inside
   is one row. A stage you are inside opens, and it opens in one of three shapes that the markup
   had already settled and nobody had written down:
     - `group`  the stage label becomes a divider and its pages become rows (User Research)
     - `nested` the same, with the pages in named groups of their own (Information Architecture,
                whose own label divider stays quiet and whose Basic / Detailed divider lights)
     - `sub`    the stage stays a link and its pages become sub-links (Wireframe Annotations)

   No em dash in this file. */

window.COURSE_ROADMAP = [
  { label: 'Foundation Research', page: 'research/research.html', dir: 'research/' },
  {
    label: 'User Research', page: 'user-research/personas.html', dir: 'user-research/', open: 'group',
    items: [
      { label: 'Personas',   page: 'user-research/personas.html' },
      { label: 'JTBD',       page: 'user-research/jtbd.html' },
      { label: 'CJM As-Is',  page: 'user-research/cjm-as-is.html' },
      { label: 'CJM To-Be',  page: 'user-research/cjm-to-be.html' }
    ]
  },
  {
    /* The only stage with two layers, and the reason is 03a and 03b: the basic layer is the map a
       person can read and the detailed layer is the one a builder can build from. The stage label
       renders as a QUIET divider here and the layer divider is the one that lights, which is what
       the 6 hand-written copies did and none of them said why. */
    label: 'Information Architecture', page: 'ia/flows.html', dir: 'ia/', open: 'nested',
    groups: [
      { label: 'Basic layer', items: [
        { label: 'Flows',       page: 'ia/flows.html' },
        { label: 'Concept map', page: 'ia/concept-map.html' }
      ] },
      { label: 'Detailed layer', items: [
        { label: 'Overview',     page: 'ia/ia.html' },
        { label: 'Sitemap',      page: 'ia/sitemap.html' },
        { label: 'SEO layer',    page: 'ia/seo.html' },
        { label: 'System nodes', page: 'ia/system.html' }
      ] }
    ]
  },
  { divider: 'Plan' },
  { label: 'Wireframes', page: 'wireframes/event-feed.html' },
  {
    label: 'Wireframe Annotations', page: 'ia/annotations/index.html', dir: 'ia/annotations/', open: 'sub',
    items: [
      { label: 'Event Feed',     page: 'ia/annotations/event-feed.html' },
      { label: 'Event Detail',   page: 'ia/annotations/event-detail.html' },
      { label: 'Category Pages', page: 'ia/annotations/category.html' },
      { label: 'My Bets',        page: 'ia/annotations/active-bets.html' },
      { label: 'Favorites',      page: 'ia/annotations/favorites.html' },
      { label: 'Deposit',        page: 'ia/annotations/deposit.html' },
      { label: 'Sign In',        page: 'ia/annotations/sign-in.html' },
      { label: 'My Profile',     page: 'ia/annotations/my-profile.html' },
      { label: 'Public Profile', page: 'ia/annotations/public-profile.html' },
      { label: 'Wallet',         page: 'ia/annotations/wallet.html' },
      { label: 'Notifications',  page: 'ia/annotations/notifications.html' },
      { label: 'How It Works',   page: 'ia/annotations/how-it-works.html' },
      { label: 'Win Screen',     page: 'ia/annotations/win.html' },
      { label: 'Loss Screen',    page: 'ia/annotations/loss.html' }
    ]
  },
  { label: 'Voice', page: 'voice/voice.html', dir: 'voice/' },
  { divider: 'Design and Delivery' },
  { label: 'Concept',       page: 'concept/concept.html', dir: 'concept/' },
  { label: 'UI + Visual',   page: 'ui-visual/event-feed.html' },
  { label: 'Design System', page: 'ui-kit/overview.html' },
  { label: 'Responsive',    page: 'ui-kit/responsive.html' },
  { label: 'Animation', planned: true },
  { label: 'Handoff',   planned: true }
];

(function () {
  /* The prefix is this file's own location, which is the only thing on the page that knows how
     deep the page is. `document.currentScript` is read at parse time on purpose: it is null by
     the time DOMContentLoaded fires. */
  var self = document.currentScript && document.currentScript.src;
  var ROOT = self ? self.replace(/assets\/_roadmap\.js.*$/, '') : '/';
  var here = location.pathname.replace(/\/index\.html$/, '/index.html');

  function isHere(page) {
    return here === '/' + page || here.slice(-(page.length + 1)) === '/' + page;
  }
  function inside(dir) {
    return dir ? here.indexOf('/' + dir) !== -1 : false;
  }
  function esc(s) { return String(s).replace(/&/g, '&amp;').replace(/</g, '&lt;'); }

  function link(item, cls) {
    var on = isHere(item.page) ? ' active' : '';
    return '<a class="' + cls + on + '" href="' + ROOT + item.page + '">' + esc(item.label) + '</a>' +
           (on ? '<!--SUB-->' : '');
  }

  /* The active stage is the DEEPEST directory that matches, because `ia/annotations/` lives
     inside `ia/` and is its own stage. A shortest-match rule would open Information Architecture
     on all fifteen annotation pages, which is the one thing the hand-written copies got right and
     a naive registry gets wrong. */
  function activeStage(list) {
    var best = null, bestLen = 0;
    list.forEach(function (s) {
      if (s.dir && inside(s.dir) && s.dir.length > bestLen) { best = s; bestLen = s.dir.length; }
      else if (!s.dir && s.page && isHere(s.page) && bestLen === 0) { best = s; }
    });
    return best;
  }

  function render() {
    var host = document.getElementById('sidebar');
    if (!host || !window.COURSE_ROADMAP) return;
    var sub = host.querySelector('[data-roadmap-sub]');
    var subHTML = sub ? sub.outerHTML.replace(' data-roadmap-sub="" hidden', '').replace(' data-roadmap-sub', '').replace(' hidden', '') : '';
    var list = window.COURSE_ROADMAP;
    var stage = activeStage(list);
    var h = '';

    /* Which divider lights is a property of the ACTIVE stage's position in the list, not of the
       divider: the one above the active stage, and only if the stage does not raise a divider of
       its own. */
    var lit = null;
    if (stage) {
      var last = null;
      for (var i = 0; i < list.length; i++) {
        if (list[i].divider) last = list[i];
        if (list[i] === stage) { lit = (stage.open === 'group' || stage.open === 'nested') ? null : last; break; }
      }
    }

    list.forEach(function (s) {
      if (s.divider) {
        h += '<div class="sidebar-divider' + (s === lit ? ' active' : '') + '">' + esc(s.divider) + '</div>';
        return;
      }
      if (s.planned) {
        h += '<span class="sidebar-page-link planned">' + esc(s.label) + '</span>';
        return;
      }
      if (s !== stage) {
        h += '<a class="sidebar-page-link" href="' + ROOT + s.page + '">' + esc(s.label) + '</a>';
        return;
      }
      if (s.open === 'group') {
        h += '<div class="sidebar-divider active">' + esc(s.label) + '</div>';
        s.items.forEach(function (i) { h += link(i, 'sidebar-page-link'); });
      } else if (s.open === 'nested') {
        h += '<div class="sidebar-divider">' + esc(s.label) + '</div>';
        s.groups.forEach(function (g) {
          var on = g.items.some(function (i) { return isHere(i.page); });
          h += '<div class="sidebar-divider' + (on ? ' active' : '') + '">' + esc(g.label) + '</div>';
          g.items.forEach(function (i) { h += link(i, 'sidebar-page-link'); });
        });
      } else if (s.open === 'sub') {
        h += '<a class="sidebar-page-link active" href="' + ROOT + s.page + '">' + esc(s.label) + '</a>';
        h += '<div class="sidebar-sub">';
        s.items.forEach(function (i) { h += link(i, 'sidebar-sub-link'); });
        h += '</div>';
      } else {
        h += '<a class="sidebar-page-link active" href="' + ROOT + s.page + '">' + esc(s.label) + '</a><!--SUB-->';
      }
    });

    /* The page's own anchors go directly under the row that is the page, and nowhere if the page
       carries none. `ia/annotations/*` has none: its sub list is route and is already drawn. */
    h = h.replace('<!--SUB-->', subHTML).replace(/<!--SUB-->/g, '');

    host.innerHTML =
      '<div class="sidebar-brand"><div class="sidebar-project-name">Prediction Market</div></div>' +
      '<nav class="sidebar-nav" aria-label="Course roadmap">' + h + '</nav>';
  }

  /* IT RENDERS NOW, NOT ON DOMContentLoaded, and the difference is not the flash. Thirteen of
     these pages carry an inline scrollspy at the foot of the body that takes
     `[...document.querySelectorAll('.sidebar-sub-link')]` ONCE, at parse time, and toggles
     `.active` on that captured list from an IntersectionObserver. A panel written after that line
     runs hands the spy a list of elements no longer in the document, and the section highlight
     dies on all thirteen with nothing in the console to say so. The script tag stands directly
     under the `<aside>` it fills, so the host is already parsed and this is simply the earliest
     correct moment. The fallback is for a page that moves the tag. */
  if (document.getElementById('sidebar')) {
    render();
  } else {
    document.addEventListener('DOMContentLoaded', render);
  }
})();
